File: Mathematics/Multiply Strings/leetcode/43. Multiply Strings/GPT/Multiply-Strings.py
# 以下は Python (CPython 3.11.4) での最適化実装です。
# 桁ごとの掛け算を配列に格納し、繰り上げを管理することで O(m×n) の計算量、O(m+n) のメモリで処理します。
# Pylance による型チェックにも対応して型ヒントを明示しています。

# ```python
from typing import List

class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        """
        文字列で与えられた2つの非負整数の積を文字列として返す

        Parameters:
            num1 (str): 掛け算する1つ目の数（文字列）
            num2 (str): 掛け算する2つ目の数（文字列）

        Returns:
            str: 積を表す文字列
        """
        # 片方が "0" の場合は即座に返す（不要な計算を避ける）
        if num1 == "0" or num2 == "0":
            return "0"

        m: int = len(num1)
        n: int = len(num2)
        # 計算結果を格納する配列（初期値は0）
        result: List[int] = [0] * (m + n)

        # 下位桁から順に掛け算（筆算と同じ要領）
        for i in range(m - 1, -1, -1):
            digit1: int = ord(num1[i]) - 48  # '0' の ASCIIコードを引く
            for j in range(n - 1, -1, -1):
                digit2: int = ord(num2[j]) - 48
                mul: int = digit1 * digit2
                sum_: int = mul + result[i + j + 1]

                result[i + j + 1] = sum_ % 10
                result[i + j] += sum_ // 10

        # 先頭のゼロをスキップ
        start: int = 0
        while start < len(result) and result[start] == 0:
            start += 1

        return "".join(map(str, result[start:]))
# ```

# ---

# ### 実装のポイント

# 1. **整数変換禁止ルール遵守**
#    → `ord(char) - 48` で桁の数値を取得（`'0'` の ASCII コードは 48）。
# 2. **O(m×n) の計算量**

#    * 最大長 200 なら 40,000 回の掛け算で十分高速。
# 3. **O(m+n) のメモリ使用**

#    * 配列 `result` は最大 401 要素で固定。
# 4. **文字列連結の最適化**

#    * 最後に `join` で一括変換し、中間的な文字列生成を避ける。

File: Mathematics/Multiply Strings/leetcode/43. Multiply Strings/GPT/Multiply-Strings.ts
// 以下は TypeScript 版の最適化実装です。
// 処理時間・メモリの消費を意識して、ループは最小限・配列は一度だけ作成しています。

// ```typescript
/**
 * 文字列として与えられた2つの非負整数の積を文字列として返す
 * @param num1 掛け算する1つ目の数（文字列）
 * @param num2 掛け算する2つ目の数（文字列）
 * @returns 積を表す文字列
 */
function multiply(num1: string, num2: string): string {
    // "0" の場合は即座に返す（不要な計算を避ける）
    if (num1 === "0" || num2 === "0") return "0";

    const m: number = num1.length;
    const n: number = num2.length;
    const result: number[] = new Array(m + n).fill(0);

    // 下位桁から掛け算
    for (let i = m - 1; i >= 0; i--) {
        const digit1: number = num1.charCodeAt(i) - 48; // '0' の ASCII を引く
        for (let j = n - 1; j >= 0; j--) {
            const digit2: number = num2.charCodeAt(j) - 48;
            const mul: number = digit1 * digit2;
            const sum: number = mul + result[i + j + 1];

            result[i + j + 1] = sum % 10;
            result[i + j] += Math.floor(sum / 10);
        }
    }

    // 先頭のゼロをスキップ
    let start: number = 0;
    while (start < result.length && result[start] === 0) {
        start++;
    }

    return result.slice(start).join("");
}
// ```

// ---

// ### 実装の特徴

// * **整数変換禁止ルールに準拠**
//   → `charCodeAt` を使い、文字→数値変換を高速に実行。
// * **O(m × n) 時間計算量**

//   * 最大長 200 なら最大 40,000 回の掛け算。
// * **メモリ効率**

//   * 計算配列 `result` は 401 要素以内で固定。
//   * 追加の文字列連結は最後に一度だけ。

